### PR TITLE
Add svd_flip (#6599)

### PR DESCRIPTION
**Pull request by eric-czech**
- [X] Tests added / passed
- [X] Passes `black dask` / `flake8 dask`

This adds the function mentioned in https://github.com/dask/dask/issues/6599.

Notes on it:

- Sign correction in SVD is a heuristic process though there seems to be some consensus that an idealized process for it identifies eigenvector directions that point in the direction of data vectors [[1](https://prod-ng.sandia.gov/techlib-noauth/access-control.cgi/2007/076422.pdf)] (this gets cited in a lot of stack exchanges / github issues)
- The "use the sign of the largest absolute value in a singular vector" approach taken in scikit-learn seems to be fairly common and is convenient when you want deterministic but arbitrary behavior (i.e. a data-independent heuristic)
- I don't think this needs to be a meaningful correction, just a deterministic one 
- The scikit-learn svd_flip function allows for the correction to be based on `U` or `V` and as best I can tell, this is only relevant when you're running SVD on transposed inputs in the first place.  I don't see any reason to need that and it appears to be used rarely so I simply left out the correct-based-on-`V` option (instead of `U`).
- The trivial dask implementation of [sklearn.svd_flip](https://github.com/scikit-learn/scikit-learn/blob/7bbd0d56530ad3758841595d81ab8deed069879d/sklearn/utils/extmath.py#L504) replaces `np.sign(u[max_abs_cols, range(u.shape[1])])` with `np.sign(u.vindex[max_abs_cols, range(u.shape[1])])` where `vindex` is used for fancy-indexing.  I found this to be unacceptably slow though.  Instead, a similarly arbitrary but more efficient correction would be to make sure all singular vectors fall in the same half-space as any arbitrarily chosen vector.  I can't find an authoritative citation for this but it's kind of a common sense approach. Here is a [notebook](https://gist.github.com/eric-czech/5c48337c034f7470328c55d93fcc730e) that compares the two approaches on equally sized short-fat or tall-skinny arrays and contains benchmark results that look like this: Screen Shot 2020-09-09 at 1 23 42 PM
    - *svd_flip1* = scikit-learn svd_flip adapted to dask arrays
    - *svd_flip2* = the implementation in this PR

Interestingly, the transposition needed to make SVD work for short-fat arrays (https://github.com/dask/dask/pull/6591) has almost no effect at all.  The scikit-learn correction using `vindex` on dask arrays almost doubles the time it takes for the whole SVD + correction routine to run but this approach using a matrix-vector multiplication instead (as a sum across rows) accounts for a fairly negligible increase in time taken.

**tl;dr** This method is efficient but I still don't know quite where to put it.  I think I would prefer if it was used based on an option in the `linalg.svd` signature set to `True` by default though that's not going to make sense until https://github.com/dask/dask/issues/3576 is done.  Also, it will mean that the setting `full_matrices=True, correct_signs=True` isn't a valid one.  For now though, I put this in `array.utils` and the test in `test_linalg.py` which is a little weird, but it seemed to make the most sense there alongside the other svd tests.

### CODE REVIEW

**Comment by eric-czech** [comment]
Another reason this could be useful is that singular vectors currently have different signs depending on the chunking:

```python
import numpy as np
import dask.array as da
rs = np.random.RandomState(1)
x = rs.random(size=(8, 4))

u, s, v = da.linalg.svd(da.asarray(x, chunks=(4, 4)))
print(u.compute().round(1))
[[ 0.3 -0.1  0.2 -0.4]
 [ 0.1  0.3 -0.3 -0.2]
 [ 0.3  0.4 -0.5  0.3]
 [ 0.4  0.4  0.5 -0.4]
 [ 0.3 -0.2 -0.1  0.3]
 [ 0.5  0.  -0.3 -0.3]
 [ 0.4 -0.7 -0.1  0.1]
 [ 0.3  0.2  0.5  0.6]]

u, s, v = da.linalg.svd(da.asarray(x, chunks=(1, 4)))
print(u.compute().round(1))
[[-0.3  0.1 -0.2 -0.4]
 [-0.1 -0.3  0.3 -0.2]
 [-0.3 -0.4  0.5  0.3]
 [-0.4 -0.4 -0.5 -0.4]
 [-0.3  0.2  0.1  0.3]
 [-0.5 -0.   0.3 -0.3]
 [-0.4  0.7  0.1  0.1]
 [-0.3 -0.2 -0.5  0.6]]
```

This would also come up with https://github.com/dask/dask/pull/6616 in chunked vs not-chunked results.

In any case, if/when this goes through I would happily integrate this in dask.svd as a part of https://github.com/dask/dask/pull/6616.  Let me know if you guys have any concerns or objections @mrocklin / @TomAugspurger.

**Comment by TomAugspurger** [comment]
Thanks for the PR and the analysis. The changes here look great.

@eric-czech when do you think a hypothetical `correct_signs=False` would be useful? Just for performance?

**Comment by eric-czech** [comment]
> @eric-czech when do you think a hypothetical correct_signs=False would be useful? Just for performance?

That's about the only good reason I can think of.  Hypothetically, it might also be useful to turn off if somebody wanted to try to support the `full_matrices=True` behavior like np.linalg.svd does since this sign correction wouldn't handle that.  The sklearn svd_flip function breaks in that case too (like we saw in https://github.com/dask/dask-ml/issues/732), but I can see how this svd_flip function could be extended to work in that scenario if it's ever necessary.

**Comment by TomAugspurger** [comment]
OK, thanks! I don't have a strong opinion, but right now my preferences align with yours (on be default, with a keyword to disable).

Let's revisit that once the other PRs and `full_matrices` issues are resolved.
